### ISSUE TITLE
Feature request - invite users + workspaces

### ISSUE THREAD

**Issue by kuba-lilz** [Flask-Middleware/flask-security#404]
A growing number of apps allows accounts to be defined per workspaces, inside which they can collaborate on shared data, instead of all accounts being fully independent entities working on independent data.

Notable examples are:
- AWS, GCP, Azure - "administrator account" + invited users accounts with permissions managed by administrator account
- Slack - "administrator account" for account that created a workspace + users invited into workspace

As far as I see Flask Security doesn't offer a mechanism to create a similar system.

I tried hacking together an "invite user" functionality, and I can sort of come close with:

```python
security = flask.current_app.extensions["security"]
datastore = security.datastore
    
kwargs = {
        "email": invited_user_email,
        "password": flask_security.utils.hash_password(invited_user_password),
}

user = datastore.create_user(**kwargs)
datastore.commit()
```

This still has a few small issues:
- user that triggers invite action has to decide passwords for invited users - yes, passwords can be reset later, but ideally invited user could decide her own password on first login
- with code above no email confirmation link is sent to invited user - so invited user has to try to log in first, to be told by app that her account is not confirmed, and request a confirmation email - though I can probably figure out how to automatically send confirmation email digging a bit deeper into Flask Security's code

There is also a second important facet to the "workspace" paradigm that services I listed above offer.
A single email account can be used for multiple accounts - one per each workspace.
E.g. `foo@bar.com` might have one account associated with `ACME CO` workspace and one account associated with `Big Evil Company`. For all practical purposes these are two separate accounts - separate entries in users table, separate API tokens, etc. They happen to share same `user.email` field, but have different `user.workspace_id` field. This poses additional challenge in logging and registration steps - we need to specify which `workspace` should logging/registration action apply to.

AWS offers per workspace registration links, e.g.:
`https://{{ workspace id }}.signin.aws.amazon.com/console`

Slack offers three fields on its login page:
- workspace name
- email
- password

So in short I would very much hope to see similar capabilities in Flask Security.
I realize that's a huge feature request that affects existing capability.

@jwag956  
Putting effort required aside - would you like to see this feature in Flask Security?
Or maybe I can already implement a similar pattern with existing tools, and I'm not digging into documentation hard enough?

**Comment by jwag956** [comment]
Thanks for your detailed and well thought out concept. I might be missing the connection - but it sure feels like 'workspaces' is a higher level concept - yes there might be some things in the User model. So on first pass - I don't see workspaces integral to FS. Adding fields to the login form and User model would seem sufficient?

The idea of being able to invite users I think is great. We have many of the pieces I think but probably need to additional glue.
First - though it isn't currently  a public API - flask_security.confirmable::send_confirmation_instructions(user) I think solves one of your issues.
Second - you might want to look into the new feature - unified signin that I recently shipped. This added a way to have an identity (email, username, phone number) and passcode (password, email link, passcode). So we have the notion of registering a user w/o a password (we store an unguessable password).

Again - although not a public API - flask_security.registerable::register_user() might be useful.

Maybe starting with those we can come up with something interesting - it might well require a slightly modified/new template for asking users to select an appropriate passcode method (reset password, set up SMS, setup authenticator, email link).

**Comment by kuba-lilz** [comment]
Thank you for the reply!

> So on first pass - I don't see workspaces integral to FS. Adding fields to the login form and User model would seem sufficient

I haven't looked into unified sign in, but if it might be a fit, I will!

> Adding fields to the login form and User model would seem sufficient?

Adding fields is one part of course, but I would still need to be able to instruct FS to use these additional fields when matching user input wit users table data on login. Maybe overriding one of `UserDatastore's` method (`find_user` perhaps?) would get me there. I suppose I need to do a bit more digging here.

I got fairy close to what I wanted with:
```python
security = application.extensions["security"]

datastore = security.datastore

kwargs = {
    "email": user_email,
    "password": flask_security.utils.hash_password(password),
    "workspace_id": workspace_id
}

user = datastore.create_user(**kwargs)
datastore.commit()

confirmation_link, _ = flask_security.confirmable.generate_confirmation_link(user)

# Send email
# Custom email sending code here
```

I essentially copied above code from `flask_security.registerable::register_user()`, skipping a few bits that handle corner cases and configuration dependent features, and replaced email sending with my own function, so I can have slightly different email template for invited users.

While with above I can invite users into a specific workspace, I can still have only one account for given email. Ideal would be to be able to have separate accounts for same email in different workspaces (like Slack or AWS), so if e.g. session cookie or API token for one workspace gets compromised, second  workspace user has access to isn't affected. Once again - I will have a look at whether overriding `UserDatastore` logic could get me there. Should have thought about that before.

**Comment by jwag956** [comment]
Ahh - I misread - I thought you were talking about a single account - so besides email do those accounts share anything else? any profile info? In other words - do you see them as separate User entries that happen to have the same email?

keep me posted on your progress - I always prefer adding something small that enables application developers to do what they need rather than implementing a complete (but possibly limited) feature. (A simple example would be to formally export some of the APIs you are using).

Also - don't forget to consider signals - there are signals for both confirmation and registration that would allow you to send emails or do any User manipulation.

Note that  once you get a user logged in with the correct account, the session will properly differentiate between them.
Also - datastore.find_user already takes multiple keys (although it doesn't if you pass in the case-insensitive option which is done for email but really doesn't need to be now that we normalize emails...).

So hurdles:
the LoginForm today doesn't have any mechanism to pass in anything but form.email.data to find_user(). You would have to override that entire login logic.

The Register form has a 'UniqueEmailFormMixin' which would not allow for multiple accounts with same email - you would need to override the register form.

And of course this would affect ChangePassword and the UnifiedSignin/Two Factor/Re-authenticate paths as well.

I do still like the idea of 'inviting' a user as a separate concept.

**Comment by kuba-lilz** [comment]
> Ahh - I misread - I thought you were talking about a single account - so besides email do those accounts share anything else? any profile info? In other words - do you see them as separate User entries that happen to have the same email?

Ideally I would go with separate User entities, and that's what Slack, AWS and others do as well.
The big advantage is that if one User entity associated with workspace A gets compromised, other workspaces aren't affected - would a big problem if compromising account with workspace A led to loss of data for organization that owns workspace B owner of compromised account happens to have access to as well.
(think e.g. of a freelance developer with access to AWS accounts of several different clients)

For hurdles:
I'm already skipping register form for invited users, so that's fine.
LoginForm and change password & friends sound harder ^^;

There's a fair chance we might go with limiting one user email to only one workspace, bypassing issues above, but if not, I'll share any progress.

Once again thank you for your comments!

**Comment by jwag956** [comment]
I have been looking into this a bit - I think a powerful enhancement that could solve part of your feature would be extend the USER_IDENTITY_ATTRIBUTES - in 4.0 this is now a list of dicts - and currently the dict can have just one entry which is the field in the User model to match. I think we can extend that so one could configure multiple fields that would be passed to the find_user method. I haven't tried to write that yet - but so far it seems like it might have multiple uses....

**Comment by kuba-lilz** [comment]
So after discussing pros and necessary changes, we decided to stay away from multiple accounts per email model for now.
We would still very much like to have it if it didn't require much additional work, but since we aren't pursuing this issue actively now, I'll close it.

Thank you for your helpful comments!
